File: algorithms/array_hashing/217-contains_duplicate.go
package main

// Time complexity: O(n)
//   - Iterate through array only once
//
// Space complexity: O(n)
//   - Have to generate a map for array
func containsDuplicate(nums []int) bool {
	nums_map := map[int]int{}
	for _, n := range nums {
		if _, ok := nums_map[n]; !ok {
			nums_map[n] = 1
		} else {
			return true
		}
	}

	return false
}

func main() {
	nums := []int{1, 2, 3, 1}
	println(containsDuplicate(nums))
}

File: algorithms/python-algorithms/array_hashing/217-contains_duplicate.py
from typing import List


# 2023-05-04
# TC: O(n) - Iterate through list of nums once
# SC: O(n) - Have to create a set DS
class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        lookup = set()

        for n in nums:
            if n in lookup:
                return True
            lookup.add(n)

        return False


def main():
    pass


if __name__ == "__main__":
    main()
